README.md
# tweepypoll

## Overview

`tweepypoll` is a Python package that allows users to extract and visualize poll data (poll questions, poll options, poll responses, etc.) from Twitter. Our goal is to make `tweepypoll` helpful and user-friendly; any Python beginner can effectively gain access to the data and make their own data-driven decisions. In particular, it could be a useful package for people doing social media journalism, or those studying social media interactions.

**NOTE**: This package assumes that the user has signed up for Twitter API Developer account, acquired the bearer token and set the environmental variable "BEARER_TOKEN". To acquire the bearer token, please follow the instructions [here](https://developer.twitter.com/en/docs/authentication/oauth-2-0/bearer-tokens).

## Functions

- `get_poll_by_id`:
    - This function extracts poll data from Twitter given the poll ID.

- `get_polls_from_user`:
    - This function retrieves a list of poll IDs from a Twitter user. These ids can be fed into the `get_poll_by_id` and `visualize_poll` functions.

- `visualize_poll`:
    - This function takes in the output of `get_poll_by_id` function and visualizes the poll information. 

## Related Packages

There are a few existing Python packages that have similar functionality for tweets from Twitter. For example, `pytweet` is a package that helps extract tweets, visualize user habit on tweet posting, and apply sentiment analysis to the data. However, there are no available packages that work specifically on polls from Twitter. 

## Installation

```bash
$ pip install tweepypoll
```

## Usage

```bash
from tweepypoll.tweepypoll import get_poll_by_id
get_poll_by_id(tweet_id)
```
**tweet_id** is numeric, such as 1481040318325739523


```bash
from tweepypoll.tweepypoll import get_polls_from_user
get_polls_from_user('user_id')
```
where **user_is** is a string username, such as 'tacobell'


```bash
from tweepypoll.tweepypoll import visualize_poll
visualize_poll(poll_obj, show_user=False, show_duration=False, show_date=False)
```
**poll_obj** is a dict outputted by get_poll_by_id(), **show_user, show_duration, show_date** are optional booleans to display username, poll duration and poll end date, respectively

## Contributors

- Wenxin Xiang
- Rada Rudyak
- Linh Giang Nguyen

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`tweepypoll` was created by Wenxin Xiang, Rada Rudyak, Linh Giang Nguyen. It is licensed under the terms of the MIT license.

## Credits

`tweepypoll` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.3.0 (2022-02-05)
### Feature
* Edit function visualize_poll ([`0fe8896`](https://github.com/UBC-MDS/tweepypoll/commit/0fe8896db3b210ef51dec6a17c10819d65a74651))
* Edit function visualize_poll ([`74ee90f`](https://github.com/UBC-MDS/tweepypoll/commit/74ee90f2d1b6266c12e1bbfe01b0135fc9f1a588))

## v0.2.3 (2022-02-01)


## v0.2.2 (2022-02-01)


## v0.2.1 (2022-02-01)


## v0.2.0 (2022-01-29)
### Feature
* Revise visualize_poll based on input dict ([`ea12cd6`](https://github.com/UBC-MDS/tweepypoll/commit/ea12cd60b9722c78f65d7b048d86f26f972d57e2))
* Revise visualize_poll ([`451448c`](https://github.com/UBC-MDS/tweepypoll/commit/451448c2c7b9d533dfeaa387fdfca3f5c7ce760f))
* Add visualize_poll ([`70b38bb`](https://github.com/UBC-MDS/tweepypoll/commit/70b38bb87a3ebbe2b08d41358c7776e7ec57846e))

## v0.1.0 (14/01/2022)

- First release of `tweepypoll`!

tests/test_tweepypoll.py
from tweepypoll.tweepypoll import visualize_poll
import pandas as pd
import altair as alt


def test_visualize_poll():
    """Test visualize_poll on a dictionary input"""

    sample_poll_obj = {
        "poll question": "What's your favorate pet?",
        "poll options": [
            {"label": "Cat", "votes": 156},
            {"label": "Dog", "votes": 256},
            {"label": "Guinea pig", "votes": 56},
        ],
        "duration": 1440,
        "date": "2019-11-28T20:26:41.000Z",
    }
    test_plot = visualize_poll(sample_poll_obj)

    # test settings on altair plot
    assert isinstance(
        test_plot, alt.Chart
    ), "The type of the argument 'poll_obj' mush be a dictionary"
    assert (
        test_plot.encoding.x.shorthand == "votes"
    ), "votes should be mapped to the x axis"
    assert (
        test_plot.encoding.y.shorthand == "label"
    ), "label should be mapped to the y axis"
    assert test_plot.mark == "bar", "mark should be a bar"
    assert (
        test_plot.encoding.color.title == "Options"
    ), "Option should be the legend title"

    # check if show_date=True, correct date and time is printed
    assert (
        pd.Timestamp(sample_poll_obj["date"]).strftime("%Y-%m-%d %H:%M:%S")
        == "2019-11-28 20:26:41"
    ), "Date and time is not correct."

    # check if show_duration=True, correct duration is printed
    assert sample_poll_obj["duration"] / 60 == 24.0, "Duration is not correct."

    # check if show_total_responses=True, correct total response is printed
    df = pd.DataFrame(sample_poll_obj["poll options"])
    assert df["votes"].sum() == 468, "Total response is not correct."


tests/test_get_poll_by_id.py
from tweepypoll.tweepypoll import get_poll_by_id
import pytest

def test_get_poll_by_id():
    """Test that function returns a dictionary"""
    rtn = get_poll_by_id([1484375486473986049])[0]
    assert isinstance(
        rtn, dict
    ), "This function must return a dictionary"

    """Test that the 'poll options' in return dict is a list"""
    opts = get_poll_by_id([1484375486473986049])[0]['poll options']
    assert isinstance(
        opts, list
    ), "Poll options is not returning a list"

    """Test that a known poll returns correct total"""
    actual = get_poll_by_id([1484375486473986049])[0]['total']
    expected = 83
    assert actual == expected, "The function is not returning correct total votes"

    """Test that function throws error when the id passed isn't numeric."""
    with pytest.raises(TypeError) as e:
        get_poll_by_id('hello')
    assert str(e.value) == 'Invalid argument type: input tweet_id must be a list of numeric IDs.'

    """Test that function notifies user if tweet does not contain a poll"""
    with pytest.raises(TypeError) as e:
        get_poll_by_id([1482319795408146433])[0]
    assert str(e.value) == 'Provided tweet does not contain a poll!'


tests/test_visualize_poll.py
from tweepypoll.tweepypoll import visualize_poll
import pandas as pd
import altair as alt


def test_visualize_poll():
    """Test visualize_poll on a dictionary input"""

    sample_poll_obj = [
        {
            "text": "Important research!!!",
            "duration": 1440,
            "date": "2022-01-22T04:01:08.000Z",
            "poll options": [
                {"position": 1, "label": "Cookies", "votes": 29},
                {"position": 2, "label": "Cupcakes", "votes": 5},
                {"position": 3, "label": "Donuts", "votes": 24},
                {"position": 4, "label": "Ice Cream", "votes": 25},
            ],
            "user": "GregShahade",
            "total": 83,
        }
    ]

    test_plot = visualize_poll(sample_poll_obj)

    # test settings on altair plot
    assert isinstance(
        test_plot[0], alt.Chart
    ), "The type of the output mush be a altair chart"
    assert (
        test_plot[0].encoding.x.shorthand == "votes"
    ), "The votes should be mapped to the x axis"
    assert (
        test_plot[0].encoding.y.shorthand == "label"
    ), "The label should be mapped to the y axis"
    assert test_plot[0].mark == "bar", "mark should be a bar"
    assert (
        test_plot[0].encoding.color.title == "Options"
    ), "Option should be the legend title"

    # check if show_user=True, correct user name is printed
    assert sample_poll_obj[0]["user"] == "GregShahade", "The user name is not correct."

    # check if show_date=True, correct date and time is printed
    assert (
        pd.Timestamp(sample_poll_obj[0]["date"]).strftime("%Y-%m-%d %H:%M:%S")
        == "2022-01-22 04:01:08"
    ), "Date and time is not correct."

    # check if show_duration=True, correct duration is printed
    assert sample_poll_obj[0]["duration"] / 60 == 24.0, "Duration is not correct."

    # check if calculated total votes is equal to the input dict
    df = pd.DataFrame(sample_poll_obj[0]["poll options"])
    assert (
        df["votes"].sum() == sample_poll_obj[0]["total"]
    ), "Total response is not correct."


src/tweepypoll/tweepypoll.py
# author: Rada Rudyak, Wenxin Xiang, Linh Giang Nguyen
# date: 2022-01-14

import altair as alt
import pandas as pd
import numpy as np
import tweepy
import dotenv
import os
import re


def get_polls_from_user(username, tweet_num=10):
    """
    Extract tweet ids (where the tweet contains a poll) for a given Twitter user

    Parameters
    ----------
    username : str
        username of the Twitter user
    tweet_num: int
        number of the most recent tweets to be pulled
        default = 10
        maximum = 100, minimum = 5 per request by Twitter API

    Returns
    --------
        array of twitter ids
    Examples
    --------
    >>> get_polls_from_user('PollzOnTwitta')
    """

    # Check argument validity
    if not (isinstance(username, str)):
        raise TypeError("Invalid argument type: username must be a string.")
    elif not (isinstance(tweet_num, int) and tweet_num >= 5 and tweet_num <= 100):
        raise TypeError("Invalid argument: input tweet_num must be >= 5 and <= 100.")

    # Twitter API credentials
    # from dotenv import load_dotenv
    # load_dotenv(".env")
    # bearer_token = os.environ.get("BEARER_TOKEN")

    ############################################################################
    # Note: For the TAs convenience, we hard coded the bearer_token below.
    # In practice, we would use commented out code to get the token from environmental variable
    ############################################################################
    bearer_token = "AAAAAAAAAAAAAAAAAAAAAAyIYQEAAAAAjvBdCMMh1dT8clkpXhHxzld7Dhs%3DLPl5zMXXOZqznZGe9JP7zHj3Wzx0N4unogLcWl8wfIkwikjQKm"
    client = tweepy.Client(bearer_token=bearer_token)

    # Get user_id from username
    users = client.get_users(usernames=username, user_fields=["id"])

    for user in users.data:
        user_id = user.id

    # Get tweets specified by the requested user id
    tweets = client.get_users_tweets(
        id=user_id, max_results=tweet_num, expansions="attachments.poll_ids"
    )

    # Get tweet_id from tweets that contain polls (if available)
    tweet_id_with_poll = []

    for tweet in tweets.data:
        if "attachments" in tweet.data.keys():
            tweet_id_with_poll.append(tweet.id)
        else:
            pass

    return tweet_id_with_poll


def get_poll_by_id(tweet_ids):
    """
    Extracts poll data from Twitter given the poll IDs array

    Parameters
    ----------
    tweet_id : str
        id of the tweet containing twitter poll

    Returns
    --------
    a list of dictionaries with the following keys:
        poll question,
        poll responses,
        total votes,
        duration,
        date

    Examples
    --------
    >>> get_poll_by_id(['1484375486473986049','1484375486473986049'])
    """

    # Check argument validity
    if not (isinstance(tweet_ids, list)):
        raise TypeError(
            "Invalid argument type: input tweet_id must be a list of numeric IDs."
        )

    # Twitter API credentials
    # from dotenv import load_dotenv, find_dotenv
    # load_dotenv(find_dotenv())
    # bearer_token = os.environ.get("BEARER_TOKEN")

    ############################################################################
    # Note: For the TAs convenience, we hard coded the bearer_token below.
    # In practice, we would use commented out code to get the token from environmental variable
    ############################################################################

    bearer_token = "AAAAAAAAAAAAAAAAAAAAAAyIYQEAAAAAjvBdCMMh1dT8clkpXhHxzld7Dhs%3DLPl5zMXXOZqznZGe9JP7zHj3Wzx0N4unogLcWl8wfIkwikjQKm"
    client = tweepy.Client(bearer_token=bearer_token)

    rtn = []

    for tweet_id in tweet_ids:

        res_tweet = client.get_tweets(
            tweet_id,
            expansions=["attachments.poll_ids", "author_id"],
            poll_fields=["duration_minutes", "end_datetime"],
        )
        res = res_tweet.includes

        try:
            res["polls"][0]
        except KeyError:
            raise TypeError("Provided tweet does not contain a poll!")

        poll = res["polls"][0]
        duration = poll["duration_minutes"]
        date = poll["end_datetime"]
        options = poll["options"]
        text = res_tweet.data[0]["text"]
        user = res["users"][0].username

        total = 0
        for opt in options:
            total = total + opt["votes"]

        tweet_data = {
            "text": text,
            "duration": duration,
            "date": date,
            "poll options": options,
            "user": user,
            "total": total,
        }

        rtn.append(tweet_data)

    return rtn


def visualize_poll(poll_obj, show_user=False, show_duration=False, show_date=False):
    """
    Returns simple bar chart(s) of poll responses
    Option to include additional information in the text box
    Parameters
    ----------
    poll_obj : list
        the output of get_poll_by_id() function
    show_user : bool
        option to display user handle in the textbox
        default = False
    show_duration : bool
        option to display poll duration
        default = False
    show_date : bool
        option to display date
        default = False
    Returns
    --------
        altair bar chart(s) for the polls responses
        includes additional information if at least one of
        - show_user
        - show_duration
        - show_date
        was set to True
    Examples
    --------
    >>> visualize_poll(tweet_polls, show_duration=True)
    """
    plot_list = []

    for i, polls in enumerate(poll_obj):

        # Check for valid inputs
        if not isinstance(poll_obj[i], dict):
            raise Exception(
                "The type of the argument 'poll_obj' mush be an array dictionary"
            )

        # convert dictionary to pd.DataFrame
        df = pd.DataFrame(poll_obj[i]["poll options"])

        # extract user id and print
        if show_user == True:
            print(f"The user of the poll: {poll_obj[i]['user']}")

        # extract poll date and print
        if show_date == True:
            print(
                f"The end date and time of the poll: {pd.Timestamp(poll_obj[i]['date']).strftime('%Y-%m-%d %H:%M:%S %Z')}"
            )

        # extract duration and print
        if show_duration == True:
            print(
                f"The duration of the poll in hours: {poll_obj[i]['duration'] / 60:.1f}h"
            )

        plot = (
            alt.Chart(
                df, title=alt.TitleParams(text=poll_obj[i]["text"], anchor="start")
            )
            .mark_bar()
            .encode(
                alt.Y("label", title="", sort="x"),
                alt.X("votes", title="Votes"),
                alt.Color("label", title="Options"),
                alt.Tooltip("votes"),
            )
            .configure_axis(
                labelFontSize=15,
                titleFontSize=15,
            )
            .configure_title(fontSize=20)
            .properties(height=200, width=400)
        )
        plot.display()
        plot_list.append(plot)
    return plot_list
